master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,248 @@
+        matchingBrids = self._getBuildRequestIdsMatchingSourceStamp(
+            builderName, priority, codebase, branch, revision)
+
+        # Check if relevant properties match
+        for otherBuildsetid, otherBrid in matchingBrids:
+            otherProperties = self.properties_cache.get(otherBuildsetid)
+            if self._propertiesMatch(properties, otherProperties,
+                                     builder.config.mergeProperties):
+                return otherBrid
+        return None
+
+    def _getBuildRequestIdsMatchingSourceStamp(self, builderName, priority,
+                                               codebase, branch, revision):
+        """
+        :param str builderName:
+        :param int priority:
+        :param str codebase:
+        :param str branch:
+        :param str revision:
+        :return tuple(str,str):
+            (buildsetid, brid) for all buildrequests that match the given
+            builderName, priority and sourceStamp information.
+
+            Note that builds started without a revision (only a branch) might
+            not be able to merge to previous builds that were also started
+            without a revision but have already started and found the latest
+            revision in that branch. This happens because in this case we will
+            be searching for other buildRequests that have `revision==None`,
+            and running builds at some point update their properties to a real
+            revision.
+        """
+        bs_tbl = self.master.db.model.buildsets
+        breq_tbl = self.master.db.model.buildrequests
+        ss_tbl = self.master.db.model.sourcestamps
+
+        def __getBuildRequestIdsMatchingSourcestamp(conn):
+            # Select buildrequests
+            # q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+
+            # That are not complete
+            #     .where(breq_tbl.c.complete == 0) \
+
+            # For this same builder
+            #     .where(breq_tbl.c.buildername == builderName) \
+
+            # That were not merged themselves (only merge against one / main target)
+            #     .where(breq_tbl.c.mergebrid == None) \
+
+            # And has the same priority (we could be smarter here, but it would make the query
+            # more complicated and possibly invert the merge target)
+            #     .where(breq_tbl.c.priority == priority) \
+
+            # Join on same sourcestamp (codebase, branch, revision)
+            #     .join(bs_tbl, bs_tbl.c.id == breq_tbl.c.buildsetid) \
+            #     .join(ss_tbl, ss_tbl.c.sourcestampsetid == bs_tbl.c.sourcestampsetid) \
+            #     .where(
+            #         ss_tbl.c.codebase == codebase,
+            #         ss_tbl.c.branch == branch,
+            #         ss_tbl.c.revision == revision,
+            #     )
+            q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+                .where(breq_tbl.c.complete == 0) \
```

Comment:
if we filter by complete builds, then we won't be merging in progress and finished builds? that's what it is in the logic of the merging function that was disabled

Reply:
Yes, and that is what we are trying to bring back (using a new code that hopefully won't slow down Katana as much as the old one)

Reply:
To make it clear, we are querying for *incomplete* builds, so we will try to merge against both pending and running builds.

Reply:
ok I think we could start by supporting only building builds merges, as for pending builds we would need to check what would be the impact of not adding a dependency row, I think as the code it is right now things could break as we used the triggeredbybrid to do things like the downloads

Reply:
I think we need to check that case anyways, merge building builds when trigger builds from dependencies.

Reply:
@campos-ddc ok ignored the previous comment, you right for _incomplete_ the check is ok